### ISSUE TITLE
Using `-o` option in combination with `-r asciidoctor-diagram` causes failure in prawn

### ISSUE THREAD

**Issue by bentolor** [asciidoctor/asciidoctor-pdf#842]
Not sure if this an issue with asciidoctor-pdf or prawn. 

When I try to generate a document with embedded UML diagrams, this fails _only_ if I use the `-o` option with an output file path.

Omitting the `-o` option works.

````
$ asciidoctor-pdf --trace -r asciidoctor-diagram -a pdf-style=doc/adoc-pdf-theme.yml doc/specification/api-specification.adoc -o /home/ben/myproject/dist/doc/api-specification.pdf
/var/lib/gems/2.3.0/gems/prawn-2.2.2/lib/prawn/document.rb:386:in `initialize': Is a directory @ rb_sysopen - /home/ben/myproject/dist/doc/api-specification.pdf (Errno::EISDIR)
	from /var/lib/gems/2.3.0/gems/prawn-2.2.2/lib/prawn/document.rb:386:in `open'
	from /var/lib/gems/2.3.0/gems/prawn-2.2.2/lib/prawn/document.rb:386:in `render_file'
	from /var/lib/gems/2.3.0/gems/asciidoctor-pdf-1.5.0.alpha.16/lib/asciidoctor-pdf/converter.rb:2828:in `write'
	from /var/lib/gems/2.3.0/gems/asciidoctor-1.5.6.1/lib/asciidoctor/document.rb:1083:in `write'
	from /var/lib/gems/2.3.0/gems/asciidoctor-1.5.6.1/lib/asciidoctor.rb:1507:in `convert'
	from /var/lib/gems/2.3.0/gems/asciidoctor-1.5.6.1/lib/asciidoctor.rb:1575:in `block in convert_file'
	from /var/lib/gems/2.3.0/gems/asciidoctor-1.5.6.1/lib/asciidoctor.rb:1575:in `open'
	from /var/lib/gems/2.3.0/gems/asciidoctor-1.5.6.1/lib/asciidoctor.rb:1575:in `convert_file'
	from /var/lib/gems/2.3.0/gems/asciidoctor-1.5.6.1/lib/asciidoctor/cli/invoker.rb:108:in `block in invoke!'
	from /var/lib/gems/2.3.0/gems/asciidoctor-1.5.6.1/lib/asciidoctor/cli/invoker.rb:102:in `each'
	from /var/lib/gems/2.3.0/gems/asciidoctor-1.5.6.1/lib/asciidoctor/cli/invoker.rb:102:in `invoke!'
	from /var/lib/gems/2.3.0/gems/asciidoctor-pdf-1.5.0.alpha.16/bin/asciidoctor-pdf:31:in `<top (required)>'
	from /usr/local/bin/asciidoctor-pdf:23:in `load'
	from /usr/local/bin/asciidoctor-pdf:23:in `<main>'
````

I can also see that inside the folder `/home/ben/myproject/dist/doc/` there exists a subfolder `api-specification.pdf` with the following content:

```
.
└── doc
    └── api-specification.pdf
        ├── .asciidoctor
        │   └── diagram
        │       ├── .domain-model.png.cache
        │       └── domain-model.png.cache
        ├── domain-model
        │   └── domain-model.png
        ├── rest-api
        │   └── .domain-model.png
        └── websocket-api
            └── .domain-model.png
```

**Comment by veeg** [comment]
Using the output option specifies which file to write the artifact (pdf) to. Since this is a directory, it fails rightfully so.

**Comment by bentolor** [comment]
> Using the output option specifies which file to write the artifact (pdf) to. 
I agree with that statement, @veeg 

> Since this is a directory, it fails rightfully so.
Yes. But this directory is not present prior to running the command. Nor does asciidoctor-pdf not fail if the document does not contain any embedded diagrams. So I do not see why it is rightfully to fail?

**Comment by mojavelinux** [comment]
Aha. Since we take over the role of writing the file in the converter, we must also make sure any directories exist. It looks like that is not being done. I would have expected Prawn to do that, but we don't have to rely on Prawn.

> Is a directory @ rb_sysopen

This is one of the oldest and most embarrassing typos in C. It reports a path in a non-existent directory as "is a directory". I have never understood why no seems to care to fix it. It should say "Is not a directory".

**Comment by bentolor** [comment]
I'm still puzzled about what the problem is. 

My original assumption was, that the generated, temporary `api-specification.pdf`-folder clashes with the final, identically named output file `api-specification.pdf`. 

Now that I learned that the error message is just wrong, I'm even more wondering what's going on…

**Comment by mojavelinux** [comment]
(Well, unless there really is a directory in the way. I haven't had a chance to test your scenario yet).

**Comment by mojavelinux** [comment]
I just checked and I was incorrect. Asciidoctor PDF does create any intermediary directories that are needed. So that's not the issue.

**Comment by mojavelinux** [comment]
This appears to be a bug in Asciidoctor Diagram (perhaps causes by a change to Asciidoctor core...but certainly a scenario that doesn't have a test). It actually has nothing to do with Asciidoctor PDF. Asciidoctor Diagram is not calculating the target directory correctly for some reason. So the issue needs to go into that project.

I'm seeing the same behavior with `asciidoctor -a data-uri`.

Btw, you can express your command using `-D /home/ben/myproject/dist/doc` instead of `-o`. But that doesn't seem to fix the problem.

**Comment by mojavelinux** [comment]
I feel confident this will be solved by the fix for #1033 (which is essentially a duplicate of this issue). If that doesn't do it, we can revisit.
